### PR TITLE
RavenDB-12890 - fix AVE during import with attachments

### PR DESCRIPTION
**Pull request by grisha-kotler**
missing failing tests

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -746,15 +756,15 @@ public void Dispose()
             }
         }
 
-        private unsafe void ReadItemsFromSource(int replicatedDocs, DocumentsOperationContext context,
+        private unsafe List<ReplicationItem> ReadItemsFromSource(int replicatedDocs, DocumentsOperationContext context,
             IncomingReplicationAllocator incomingReplicationAllocator, IncomingReplicationStatsScope stats)
         {
+            var replicatedItems = new List<ReplicationItem>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: Can we avoid this allocation? Return `Span<ReplicationItem>` and user `ArrayPool`, instead?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 2]
```diff
@@ -994,9 +1006,9 @@ public void Dispose()
             }
         }
 
-        private unsafe void ReadAttachmentStreamsFromSource(int attachmentStreamCount, DocumentsOperationContext context)
+        private unsafe Dictionary<Slice, ReplicationAttachmentStream> ReadAttachmentStreamsFromSource(int attachmentStreamCount, DocumentsOperationContext context)
         {
-            Debug.Assert(_replicatedAttachmentStreams.Count == 0, "We should handle all attachment streams during WriteAttachment.");
+            var replicatedAttachmentStreams = new Dictionary<Slice, ReplicationAttachmentStream>(SliceComparer.Instance);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1034]: Same, can we reuse?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 3]
```diff
@@ -517,102 +498,158 @@ private unsafe void ReadExactly(byte* ptr, int size)
...
+            internal Dictionary<Slice, ReplicationAttachmentStream> ReplicatedAttachmentStreams { get; set; }
+
+            public TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures { get; set; }
+
+            public Logger Logger { get; set; }
+
+            public void Dispose()
             {
-                using (var networkStats = stats.For(ReplicationOperation.Incoming.Network))
+                if (ReplicatedItems != null)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: ReplicatedItems is a struct, can never be null. You probably want just iterate over it.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 4]
```diff
@@ -517,102 +498,158 @@ private unsafe void ReadExactly(byte* ptr, int size)
...
                 {
-                    // this will read the documents to memory from the network
-                    // without holding the write tx open
-                    ReadItemsFromSource(replicatedItemsCount, documentsContext, incomingReplicationAllocator, networkStats);
-
-                    using (networkStats.For(ReplicationOperation.Incoming.AttachmentRead))
+                    foreach (var item in ReplicatedItems)
                     {
-                        ReadAttachmentStreamsFromSource(attachmentStreamCount, documentsContext);
+                        item.Document.Dispose();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]:
```suggestion
                        item?.Document?.Dispose();
```

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 5]
```diff
@@ -517,102 +498,158 @@ private unsafe void ReadExactly(byte* ptr, int size)
...
 
-                if (_log.IsInfoEnabled)
-                {
-                    _log.Info(
-                        $"Replication connection {FromToString}: " +
-                        $"received {replicatedItemsCount:#,#;;0} items, " +
-                        $"{attachmentStreamCount:#,#;;0} attachment streams, " +
-                        $"total size: {new Size(incomingReplicationAllocator.TotalDocumentsSizeInBytes, SizeUnit.Bytes)}, " +
-                        $"took: {sw.ElapsedMilliseconds:#,#;;0}ms");
+                    if (_isReplayTransaction == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: Why do we need to make this distinction? It it better if the caller always pass an array from the array pool

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 6]
```diff
@@ -1500,30 +1532,37 @@ public IncomingReplicationHandler.MergedUpdateDatabaseChangeVectorCommand ToComm
...
         public ReplicationItemDto[] ReplicatedItemDtos;
         public long LastEtag;
+        public TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures;
         public string SourceDatabaseId;
         public KeyValuePair<string, Stream>[] ReplicatedAttachmentStreams;
 
         public IncomingReplicationHandler.MergedDocumentReplicationCommand ToCommand(DocumentsOperationContext context, DocumentDatabase database)
         {
-            var connectionInfo = new IncomingConnectionInfo
+            var replicationItems = ReplicatedItemDtos.Select(d => d.ToItem(context)).ToArray();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: Get this array from the pool

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: item cannot be null

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: but document can be

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 6 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 7]
```diff
@@ -517,102 +498,151 @@ private unsafe void ReadExactly(byte* ptr, int size)
...
+
+            internal Dictionary<Slice, ReplicationAttachmentStream> ReplicatedAttachmentStreams { get; set; }
+
+            public TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures { get; set; }
+
+            public Logger Logger { get; set; }
+
+            public void Dispose()
+            {
+                foreach (var item in ReplicatedItems)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: Check if ReplicatedItems are not `null`, when you will get exception in `ReadItemsFromSource` then you will get NRE here

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 8]
```diff
@@ -814,6 +891,155 @@ public MergedBatchPutCommand ToCommand(DocumentsOperationContext context, Docume
...
+            {
+                if (_log.IsInfoEnabled)
+                    _log.Info($"Trying to update {Ids.Count:#,#0} documents metadata if necessary");
+
+                var count = 0;
+                foreach (var id in Ids)
+                {
+                    using (DocumentIdWorker.GetSliceFromId(context, id, out Slice lowerId))
+                    {
+                        if (_database.DocumentsStorage.GetTableValueReaderForDocument(context, lowerId, throwOnConflict: true, out var tvr) == false)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: This looks like `DocumentsStorage.Get`. We should not expose the method externally? Also probably you do not want to throw on conflict?

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 9]
```diff
@@ -1046,7 +1046,7 @@ private static Document TableValueToDocument(DocumentsOperationContext context,
         }        
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static Document ParseDocument(JsonOperationContext context, ref TableValueReader tvr)
+        public static Document ParseDocument(JsonOperationContext context, ref TableValueReader tvr)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: Can be private after using `Get` in the command

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 10]
```diff
@@ -1500,30 +1528,56 @@ public IncomingReplicationHandler.MergedUpdateDatabaseChangeVectorCommand ToComm
...
+            var dataForReplicationCommand = new IncomingReplicationHandler.DataForReplicationCommand()
+            {
+                DocumentDatabase = database,
+                ConflictManager = new ConflictManager(database, database.ReplicationLoader.ConflictResolver),
+                SourceDatabaseId = SourceDatabaseId,
+                ReplicatedItems = new ArraySegment<IncomingReplicationHandler.ReplicationItem>(replicationItems, 0, replicatedItemsCount),
+                AttachmentStreams = attachmentStreams,
+                ReplicatedAttachmentStreams = replicatedAttachmentStreams,
+                SupportedFeatures = SupportedFeatures,
+                Logger = LoggingSource.Instance.GetLogger<IncomingReplicationHandler>(database.Name)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1576]: why we are creating logger each time?

**Review by ppekrol — APPROVED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**